libp2p-identify/src/transport.rs

```diff
@@ -0,0 +1,384 @@
+#[derive(Debug, Clone)]
+pub struct IdentifyTransport<T, P> {
+	transport: T,
+	peerstore: P,
+	addr_ttl: Duration,
+}
+
+impl<T, P> IdentifyTransport<T, P> {
+	/// Creates an `IdentifyTransport` that wraps around the given transport and peerstore.
+	#[inline]
+	pub fn new(transport: T, peerstore: P) -> IdentifyTransport<T, P> {
+		IdentifyTransport::with_ttl(transport, peerstore, Duration::from_secs(3600))
+	}
+
+	/// Same as `new`, but allows specifying a time-to-live for the addresses gathered from
+	/// remotes that connect to us.
+	///
+	/// The default value is one hour.
+	#[inline]
+	pub fn with_ttl(transport: T, peerstore: P, ttl: Duration) -> IdentifyTransport<T, P> {
+		IdentifyTransport {
+			transport: transport,
+			peerstore: peerstore,
+			addr_ttl: ttl,
+		}
+	}
+}
+
+impl<T, Pr, P> Transport for IdentifyTransport<T, Pr>
+where
+	T: Transport + Clone + 'static,          // TODO: 'static :(
+	Pr: Deref<Target = P> + Clone + 'static, // TODO: 'static :(
+	for<'r> &'r P: Peerstore,
+{
+	type RawConn = T::RawConn;
+	type Listener = Box<Stream<Item = Self::ListenerUpgrade, Error = IoError>>;
+	type ListenerUpgrade = Box<Future<Item = (T::RawConn, Multiaddr), Error = IoError>>;
+	type Dial = Box<Future<Item = (T::RawConn, Multiaddr), Error = IoError>>;
+
+	#[inline]
+	fn listen_on(self, addr: Multiaddr) -> Result<(Self::Listener, Multiaddr), (Self, Multiaddr)> {
+		// Note that `listen_on` expects a "regular" multiaddr (eg. `/ip/.../tcp/...`),
+		// and not `/p2p/<foo>`.
+
+		let (listener, new_addr) = match self.transport.clone().listen_on(addr.clone()) {
+			Ok((l, a)) => (l, a),
+			Err((inner, addr)) => {
+				let id = IdentifyTransport {
+					transport: inner,
+					peerstore: self.peerstore,
+					addr_ttl: self.addr_ttl,
+				};
+				return Err((id, addr));
+			}
+		};
+
+		let identify_upgrade = self.transport.with_upgrade(IdentifyProtocolConfig);
+		let peerstore = self.peerstore;
+		let addr_ttl = self.addr_ttl;
+
+		let listener = listener.map(move |connec| {
+			let peerstore = peerstore.clone();
+			let identify_upgrade = identify_upgrade.clone();
+			let fut = connec
+				.and_then(move |(connec, client_addr)| {
+					// Dial the address that connected to us and try upgrade with the
+					// identify protocol.
+					identify_upgrade
+						.clone()
+						.dial(client_addr.clone())
+						.map_err(|_| {
+							IoError::new(IoErrorKind::Other, "couldn't dial back incoming node")
+						})
+						.map(move |id| (id, connec))
+				})
+				.and_then(move |(dial, connec)| dial.map(move |dial| (dial, connec)))
+				.and_then(move |((identify, original_addr), connec)| {
+					// Compute the "real" address of the node (in the form `/p2p/...`) and add
+					// it to the peerstore.
+					let real_addr = match identify {
+						IdentifyOutput::RemoteInfo { info, .. } => process_identify_info(
+							&info,
+							&*peerstore.clone(),
+							original_addr,
+							addr_ttl,
+						)?,
+						_ => unreachable!(
```

Comment:
I'd like if we allowed protocols to define that they only work on servers or clients or both so that we could have a seperate `IdentifyServer` and `IdentifyClient` type and didn't need hacks like this, but that's a major change and not necessary for this PR.